Mathlib/Algebra/Group/Defs.lean

```diff
@@ -620,11 +620,134 @@ the file `Algebra.GroupPower.Basic`. For now, we only register the most basic pr
+  go k 1 m
+where
+  /-- Auxiliary tail-recursive implementation for `npowBinRec`-/
+  go : ℕ → M → M → M
+  | 0, y, _ => y
+  | (k + 1), y, x =>
+    let k' := (k + 1) >>> 1
+    if k &&& 1 = 1 then
+      have : k' < k + 1 := Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+      go k' y (x * x)
+    else
+      have : k' < k + 1 := Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+      go k' (y * x) (x * x)
+  termination_by go k _ _ => k
+
+attribute [to_additive existing] npowBinRec.go
+
+-- The unusedHavesSuffices linter is a false positive, see https://github.com/leanprover/std4/issues/428
+@[to_additive (attr := nolint unusedHavesSuffices)]
+theorem npowBinRec.go_spec {M : Type*} [Semigroup M] [One M]
+    (mul_one : ∀ m : M, m * 1 = m) (one_mul : ∀ m : M, 1 * m = m) (k : ℕ) (x y : M) :
+    npowBinRec.go k y x = y * npowRec' mul_one one_mul k x := by
+  induction k using Nat.strongInductionOn generalizing x y with
+  | ind k' ih =>
+    cases k' with
+    | zero => rw [go, npowRec', npowRec, mul_one]
+    | succ k' =>
+      rw [go]
+      dsimp
+      split <;> rename_i w
+      · rw [ih]
+        · congr 1
+          rw [← npowRec'_two_mul]
+          congr 1
+          rw [Nat.shiftRight_eq_div_pow, Nat.pow_one, Nat.mul_div_cancel']
+          simp only [Nat.and_one_is_mod] at w
+          rw [Nat.dvd_iff_mod_eq_zero, Nat.add_mod, w]
+          decide
+        · exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+      · rw [ih]
+        · rw [mul_assoc y x]
+          rw [← npowRec'_two_mul]
+          congr 1
+          conv =>
+            rhs
+            rw [npowRec', npowRec]
+          congr 2
+          rw [Nat.shiftRight_eq_div_pow, Nat.pow_one]
+          simp only [Nat.and_one_is_mod] at w
+          replace w : k' % 2 = 0 := by
+            cases (Nat.mod_two_eq_zero_or_one k')
+            · assumption
+            · contradiction
+          conv =>
+            lhs
+            rw [← Nat.div_add_mod k' 2]
+          rw [w, Nat.add_zero, Nat.mul_add_div (by decide), show 1 / 2 = 0 by decide, Nat.add_zero,
+            Nat.mul_div_cancel']
+          rwa [Nat.dvd_iff_mod_eq_zero]
+        · exact Nat.div_lt_self (Nat.zero_lt_succ _) (Nat.lt_succ_self _)
+
+@[csimp] theorem nsmulRec_eq_nsmulBinRec : @nsmulRec' = @nsmulBinRec := by
+  funext M _ _ add_zero zero_add k m
+  dsimp [nsmulBinRec]
+  rw [nsmulBinRec.go_spec add_zero, zero_add]
+
+@[csimp] theorem npowRec_eq_npowBinRec : @npowRec' = @npowBinRec := by
+  funext M _ _ mul_one one_mul k m
+  dsimp [npowBinRec]
+  rw [npowBinRec.go_spec mul_one, one_mul]
 
 /-- An `AddMonoid` is an `AddSemigroup` with an element `0` such that `0 + a = a + 0 = a`. -/
 class AddMonoid (M : Type u) extends AddSemigroup M, AddZeroClass M where
   /-- Multiplication by a natural number. -/
-  protected nsmul : ℕ → M → M := nsmulRec
+  protected nsmul : ℕ → M → M := @nsmulRec' M _ _ add_zero zero_add
```

Comment:
It doesn't seem like there is much conflict? If you want to remove the default, just advise users to use this new default instead.

(If being able to omit the arguments `mul_one` / `one_mul` is important, we could interpose a `PreMonoid` class that does the `extends Semigroup, MulOneClass`, and then `PreMonoid` adds the `npow*` fields; in this PR I decide that the slightly verbosity of passing the unpacked fields was a win relative to the possible performance cost of another layer.)

Similarly, the new `npowBinRec` and associated proofs could be trivially refactored to only use power associativity rather than associativity.

Reply:
The problem is that you can't supply the `add_zero` and `zero_add` arguments if the user fills the field explicitly; self-references are only allowed in default values.



> (We could provide another wrapper around `npowRec'` that has `Semigroup` / `MulOneClass` hypotheses)

This unfortunately doesn't work, as then we get `Mul` twice. We could have a `Monoid.WithoutPowYet` typeclass that extends both, but that's obviously pretty ugly...

Reply:
I'm confused by what the objection is here, to be honest. #6262 proposes removing default values, and instead writing them in concrete instances; can't it do that equally well before or after this PR?

Reply:
@eric-wieser, I agree with Scott. I don't see what the problem is. Can you be more explicit?

Reply:
(I'd really like to be able to proceed with this change, as it enables the demo that @robertylewis and I were working on with some CAS developers at a recent workshop. Shipping the demo is a chance to advertise Lean's capabilities to the CAS world.)

Reply:
Sorry, let me be more explicit:

On master, we can write
```
instance : Monoid Foo where
  one := sorry
  mul := sorry
  one_mul := sorry
  mul_one := sorry
  mul_assoc := sorry
```
With #6262 (or at least, the `pow` version of it), we can no longer write that, because if we add `npow := npowRec` then Lean can't find the `1` or `*` in time. So we have to write
```lean
instance : Mul Foo where
  mul := sorry

instance : One Foo where
  one := sorry

instance : Monoid Foo where
  one_mul := sorry
  mul_one := sorry
  mul_assoc := sorry
  npow := npowRec
```
with #6262 + this change, things get even worse, because we need to set up all the previous smaller structures before we can call `npowRec'`
```lean
instance : Mul Foo where
  mul := sorry

instance : One Foo where
  one := sorry

instance : MulOneClass Foo where
  one_mul := sorry
  mul_one := sorry

instance : Semigroup Foo where
  mul_assoc := sorry

instance : Monoid Foo where
  npow := npowRec mul_one one_mul -- slightly more annoying
```

Essentially the problem is that default arguments have magic powers that regular arguments cannot attain, and your proposal uses more of those magical powers.

Note that my objection is only to setting the default power operation; I'm very happy for it to be manually overriden on a handful of types to pique the interest of the CAS people.

Reply:
Okay. My preference is still to proceed with this PR. If the decision is not too, I can eventually make this change just on Int/Nat/ZMod/etc, but not for at least a few weeks.

Reply:
Given #12608, I am resolving this thread. Please let me know if I am mistaken.